### PR TITLE
Add OptionalUtils

### PR DESCRIPTION
**Pull request by dawidxc**
Move general-purpose methods related to `Optional` from `TeradataUtils` to `OptionalUtils`. Use the methods to simplify code.

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/ConnectorArguments.java [hunk 1]
```diff
@@ -976,11 +975,7 @@ public String toString() {
 
   @CheckForNull
   public String getDefinitionOrDefault(ConnectorPropertyWithDefault property) {
-    String stringValue = getDefinition(property);
-    if (StringUtils.isEmpty(stringValue)) {
-      return property.getDefaultValue();
-    }
-    return stringValue;
+    return nonEmpty(getDefinition(property)).orElseGet(property::getDefaultValue);
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/ConnectorArguments.java:978]: Are we sure this is easier to read?

Comment (dawidxc) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/ConnectorArguments.java]: It depends. This one is less important compared to other changes in this PR, so I've reverted it.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package com.google.edwmigration.dumper.application.dumper.utils;
+
+import com.google.common.base.Strings;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class OptionalUtils {
+
+  public static <T> Optional<T> optionalIf(boolean condition, Supplier<T> supplier) {
```
Comment (cyulysses-corp) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java:25]: nit: the meaning of the name is not clear to me. I would recommend something like:

optionallyNonEmptyIf

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return condition ? Optional.of(supplier.get()) : Optional.empty();
+  }
+
+  public static Optional<String> nonEmpty(String value) {
```
Comment (cyulysses-corp) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java:29]: nit: optionallyNullableIfEmpty

**Review by cyulysses-corp — APPROVED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java (hunk 2 above)
Comment (dawidxc) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/OptionalUtils.java:25]: Renamed to `optionallyWhen` in #359 as discussed offline.

**Comment by shevek-google** [comment]
My overall suggestion on PRs like this is that Optional does not improve the readability of the code, and readability is a critical requirement for code which is to be externally audited. I suspect the nullness analyzer in Checker Framework is better at doing the dataflow analysis than the (arguably technically equivalent, but nevertheless unimplemented) work to make sure isPresent() is always called before get(). We could respond "But we never call get, we always call orElse()" but that doesn't have the same assertion strength as Objects.requireNonNullElse, which also asserts on the second argument.

**Pull request merged**